WireDomain/Sources/WireDomain/Repositories/FeatureConfigRepository.swift

```diff
@@ -0,0 +1,363 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+import Combine
+import WireAPI
+import WireDataModel
+
+/// Facilitates access to feature configs related domain objects.
+
+protocol FeatureConfigRepositoryProtocol {
+
+    /// Pulls feature configs from the server and stores them locally.
+    ///
+
+    func pullFeatureConfigs() async throws
+
+    /// Observes feature states.
+    ///
+    /// Each time `pullFeatureConfigs()` is called, a feature config is
+    /// stored locally and a new `FeatureState` value is produced by the publisher.
+    /// It allows the user to be notified of any feature changes over time.
+    ///
+    /// - Warning:  Use this method before calling `pullFeatureConfigs` to receive all emitted values.
+    ///
+    /// - Returns: A publisher of `FeatureState`.
+
+    func observeFeatureStates() -> AnyPublisher<FeatureState, Never>
+
+    /// Fetches a feature config locally.
+    ///
+    /// - Parameter name: The feature name to fetch the config for.
+    /// - Parameter type: The type of config to retrieve.
+    /// - Returns: A `LocalFeature` object with a status and a config (if any).
+
+    func fetchFeatureConfig<T: Decodable>(withName name: Feature.Name, type: T.Type) async throws -> LocalFeature<T>
+
+    /// Fetches a flag indicating whether the user should be notified of a given feature.
+    /// - Parameter name: The feature name.
+    /// - Returns: `true` if user should be notified.
+
+    func fetchNeedsToNotifyUser(forFeatureName name: Feature.Name) async throws -> Bool
```

Comment:
I don't see what type is Feature.Name but if it's a concrete type we can shorten the signature
```suggestion
    func fetchNeedsToNotifyUser(for name: Feature.Name) async throws -> Bool
```

Reply:
it's an enum, I feel like having an explicit parameter name gives more clarity at call site:

`fetchNeedsToNotifyUser(forFeatureName: .fileSharing)` rather than `fetchNeedsToNotifyUser(for: .fileSharing)` . Looking at the second version we don't right away know what is fileSharing. I'm ok doing both just thinking out loud here

Reply:
my rule would be if you have a string, it makes sense to be explicit as you don't know what it represents.
but in case of an enum you will know what it represents because its type give you the information

See https://www.swift.org/documentation/api-design-guidelines/#conventions

Reply:
fixed